Starting node on new hf1 network with lfs would get inconsistent dag error

<!-- Please answer these questions before submitting a bug report. -->
docker run command
```
        "run",
        "--bootstrap=rnode://ef34d5e196c8292cfe0da69d7df2e9ae6cea0be2@node0.root-shard.mainnet.rchain.coop?protocol=40400&discovery=40404",
        "--network-id=mainnet1",
        "--shard-name=root1",
        "--fault-tolerance-threshold=-1",
        "--api-enable-reporting",
        "--finalization-rate=1",
        "--max-number-of-parents=1"
```


After the node finish downloading the state and the approved block, the  node would raise error like below
```
rnode    | 05:26:01.590 [INFO ] [node-runner-28      ] [c.r.c.engine.Initializing    ] - Approved state for block #908739 (f618d141a5...) is successfully restored.
rnode    | 05:26:02.239 [INFO ] [node-runner-29      ] [c.r.c.MultiParentCasperImpl  ] - Validating block #908740 (d55b77c98f...).
rnode    | 05:26:02.654 [ERROR] [node-runner-28      ] [c.r.node.runtime.NodeRuntime ] - Caught unhandable error. Exiting. Stacktrace below.
rnode    | coop.rchain.blockstorage.dag.BlockDagInconsistencyError: DAG storage is missing hash b9df500556...
rnode    |      at coop.rchain.blockstorage.dag.BlockDagRepresentationOps$.$anonfun$lookupUnsafe$extension0$1(BlockDagRepresentationSyntax.scala:41)
rnode    |      at cats.ApplicativeError$LiftFromOptionPartially$.apply$extension(ApplicativeError.scala:305)
rnode    |      at cats.syntax.OptionOps$LiftToPartiallyApplied.apply(option.scala:374)
rnode    |      at coop.rchain.blockstorage.dag.BlockDagRepresentationOps$.$anonfun$lookupUnsafe$1(BlockDagRepresentationSyntax.scala:41)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |      at runToFuture @ coop.rchain.catscontrib.TaskContrib$TaskOps$.unsafeRunSync$extension(taskOps.scala:27)
rnode    |
rnode exited with code 0
```

--- a/casper/src/main/scala/coop/rchain/casper/engine/BlockRetriever.scala
+++ b/casper/src/main/scala/coop/rchain/casper/engine/BlockRetriever.scala
@@ -198,7 +198,7 @@ object BlockRetriever {
           _ <- result.status match {
                 case NewSourcePeerAddedToRequest =>
                   Log[F]
-                    .info(
+                    .debug(
                       s"Adding ${peer.get.endpoint.host} to waiting list of ${PrettyPrinter
                         .buildString(hash)} request. " +
                         s"Reason: $admitHashReason"

--- a/casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
+++ b/casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
@@ -223,8 +223,26 @@ class Initializing[F[_]
     def getMinBlockHeight: F[Long] = {
       val blockHashes = startBlock.justifications.map(_.latestBlockHash)
       for {
-        blocks    <- blockHashes.traverse(BlockStore[F].getUnsafe)
-        minHeight = if (blocks.isEmpty) 0L else blocks.map(ProtoUtil.blockNumber).min
+        lowestBlockHeights <- fs2.Stream
+                               .fromIterator(blockHashes.toIterator)
+                               .evalMap { BlockStore[F].getUnsafe(_) }
+                               .evalMap { b =>
+                                 val noSelfJustificationErr =
+                                   "No latest messages justifications downloaded when reconstructed Last Finalized State"
+                                 val selfJ =
+                                   ProtoUtil
+                                     .creatorJustification(b)
+                                     .map(_.latestBlockHash)
+                                     .get
+                                 BlockStore[F]
+                                   .get(selfJ)
+                                   .flatMap(_.liftTo(new Exception(noSelfJustificationErr)))
+                                   .map(_.body.state.blockNumber)
+                               }
+                               .compile
+                               .toList
+
+        minHeight = if (lowestBlockHeights.isEmpty) 0L else lowestBlockHeights.min
         // We need parent of oldest latest block, that's why -1.
         // The same condition is also applied when downloading.
         // https://github.com/rchain/rchain/blob/0a09467628/casper/src/main/scala/coop/rchain/casper/engine/LastFinalizedStateBlockRequester.scala#L154

--- a/casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
+++ b/casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala
@@ -193,7 +193,7 @@ class Initializing[F[_]
 
       // Receive the blocks and after populate the DAG
       blockRequestAddDagStream = blockRequestStream.last.unNoneTerminate.evalMap { st =>
-        populateDag(approvedBlock.candidate.block, minBlockNumberForDeployLifespan, st.heightMap)
+        populateDag(approvedBlock.candidate.block, st.lowerBound, st.heightMap)
       }
 
       // Run both streams in parallel until tuple space and all needed blocks are received
@@ -215,39 +215,10 @@ class Initializing[F[_]
 
   private def populateDag(
       startBlock: BlockMessage,
-      initialMinHeight: Long,
+      minHeight: Long,
       heightMap: SortedMap[Long, Set[BlockHash]]
   ): F[Unit] = {
     import cats.instances.list._
-
-    def getMinBlockHeight: F[Long] = {
-      val blockHashes = startBlock.justifications.map(_.latestBlockHash)
-      for {
-        lowestBlockHeights <- fs2.Stream
-                               .fromIterator(blockHashes.toIterator)
-                               .evalMap { BlockStore[F].getUnsafe(_) }
-                               .evalMap { b =>
-                                 val noSelfJustificationErr =
-                                   "No latest messages justifications downloaded when reconstructed Last Finalized State"
-                                 val selfJ =
-                                   ProtoUtil
-                                     .creatorJustification(b)
-                                     .map(_.latestBlockHash)
-                                     .get
-                                 BlockStore[F]
-                                   .get(selfJ)
-                                   .flatMap(_.liftTo(new Exception(noSelfJustificationErr)))
-                                   .map(_.body.state.blockNumber)
-                               }
-                               .compile
-                               .toList
-
-        minHeight = if (lowestBlockHeights.isEmpty) 0L else lowestBlockHeights.min
-        // We need parent of oldest latest block, that's why -1.
-        // The same condition is also applied when downloading.
-        // https://github.com/rchain/rchain/blob/0a09467628/casper/src/main/scala/coop/rchain/casper/engine/LastFinalizedStateBlockRequester.scala#L154
-      } yield Math.min(minHeight - 1, initialMinHeight)
-    }
 
     def addBlockToDag(block: BlockMessage, isInvalid: Boolean): F[Unit] =
       Log[F].info(
@@ -265,7 +236,6 @@ class Initializing[F[_]
         .toSet
 
       // Add sorted DAG in order from approved block to oldest
-      minHeight <- getMinBlockHeight
       _ <- heightMap.flatMap(_._2).toList.reverse.traverse_ { hash =>
             for {
               block <- BlockStore[F].getUnsafe(hash)

--- a/casper/src/main/scala/coop/rchain/casper/engine/LfsBlockRequester.scala
+++ b/casper/src/main/scala/coop/rchain/casper/engine/LfsBlockRequester.scala
@@ -50,7 +50,13 @@ object LfsBlockRequester {
     //  in case of resend together with Requested.
     // Returns updated state with requested keys.
     def getNext(resend: Boolean): (ST[Key], Set[Key]) = {
-      val requested = d
+      val requests = {
+        // Check is latest are requested
+        if (latest.isEmpty) d
+        // Add latest to Init if not already
+        else d ++ (latest -- d.keySet).map((_, Init))
+      }
+      val newRequests = requests
         .filter {
           case (key, status) =>
             // Select initialized or re-request if resending
@@ -64,7 +70,7 @@ object LfsBlockRequester {
             }
         }
         .mapValues(_ => Requested)
-      this.copy(d ++ requested) -> requested.keySet
+      this.copy(d ++ newRequests) -> newRequests.keySet
     }
 
     // Confirm key is Received if it was Requested.
@@ -77,9 +83,10 @@ object LfsBlockRequester {
       val isReq = d.get(k).contains(Requested)
       if (isReq) {
         // Remove message from the set of latest messages (if exists)
-        val adjLatest    = latest - k
+        val adjLatest = latest - k
+        val isLatest  = latest != adjLatest
+        // Add replacement message if supplied
         val newLatest    = latestReplacement.map(adjLatest + _).getOrElse(adjLatest)
-        val isLatest     = latest != newLatest
         val isLastLatest = isLatest && newLatest.isEmpty
         // Save in height map
         val heightKeys   = heightMap.getOrElse(height, Set())
@@ -87,8 +94,7 @@ object LfsBlockRequester {
         // Calculate new minimum height if latest message
         //  - we need parents of latest message so it's `-1`
         val newLowerBound = if (isLatest) Math.min(height - 1, lowerBound) else lowerBound
-        val newSt =
-          latestReplacement.foldLeft(d + ((k, Received)))((acc, k) => acc + (k -> Requested))
+        val newSt         = d + ((k, Received))
         // Set new minimum height and update latest
         (
           this.copy(newSt, newLatest, newLowerBound, newHeightMap),
@@ -161,7 +167,7 @@ object LfsBlockRequester {
     // - for approved state to be complete it is required to have block from each of them
     val latestMessages = block.justifications.map(_.latestBlockHash).toSet
 
-    val initialHashes = latestMessages + block.blockHash
+    val initialHashes = Set(block.blockHash)
 
     def createStream(
         st: Ref[F, ST[BlockHash]],
@@ -326,7 +332,7 @@ object LfsBlockRequester {
       st <- Ref.of[F, ST[BlockHash]](
              ST(
                initialHashes,
-               latest = initialHashes,
+               latest = latestMessages,
                lowerBound = initialMinimumHeight
              )
            )

--- a/casper/src/main/scala/coop/rchain/casper/engine/LfsBlockRequester.scala
+++ b/casper/src/main/scala/coop/rchain/casper/engine/LfsBlockRequester.scala
@@ -69,11 +69,16 @@ object LfsBlockRequester {
 
     // Confirm key is Received if it was Requested.
     // Returns updated state with the flags if Requested and last latest received.
-    def received(k: Key, height: Long): (ST[Key], ReceiveInfo) = {
+    def received(
+        k: Key,
+        height: Long,
+        latestReplacement: Option[Key] = None
+    ): (ST[Key], ReceiveInfo) = {
       val isReq = d.get(k).contains(Requested)
       if (isReq) {
         // Remove message from the set of latest messages (if exists)
-        val newLatest    = latest - k
+        val adjLatest    = latest - k
+        val newLatest    = latestReplacement.map(adjLatest + _).getOrElse(adjLatest)
         val isLatest     = latest != newLatest
         val isLastLatest = isLatest && newLatest.isEmpty
         // Save in height map
@@ -82,7 +87,8 @@ object LfsBlockRequester {
         // Calculate new minimum height if latest message
         //  - we need parents of latest message so it's `-1`
         val newLowerBound = if (isLatest) Math.min(height - 1, lowerBound) else lowerBound
-        val newSt         = d + ((k, Received))
+        val newSt =
+          latestReplacement.foldLeft(d + ((k, Received)))((acc, k) => acc + (k -> Requested))
         // Set new minimum height and update latest
         (
           this.copy(newSt, newLatest, newLowerBound, newHeightMap),
@@ -192,7 +198,15 @@ object LfsBlockRequester {
         val blockNumber = ProtoUtil.blockNumber(block)
         for {
           // Mark block as received and calculate minimum height (if latest)
-          receivedResult <- st.modify(_.received(block.blockHash, blockNumber))
+          receivedResult <- st.modify(st => {
+                             // if message received is latest as per approved block - add its self justification
+                             // to target latest messages that has to be pulled
+                             val lmReplacement =
+                               if (latestMessages.contains(block.blockHash))
+                                 ProtoUtil.creatorJustification(block).map(_.latestBlockHash)
+                               else None
+                             st.received(block.blockHash, blockNumber, lmReplacement)
+                           })
           // Result if block is received and if last latest is received
           ReceiveInfo(isReceived, isReceivedLatest, isLastLatest) = receivedResult
 

